### PR TITLE
feat(toast): L3-6424 Add Toast Component

### PR DESCRIPTION
**Pull request by smarks8**
**Jira ticket**

[L3-6424](https://phillipsauctions.atlassian.net/browse/L3-6424)

**Screenshots**
image

**Figma link**

[Figma Link](https://www.figma.com/design/f3vm6hItGjnsmkJxLLbe3P/MANAGE---MUI-v5.9.0--24?node-id=6586-47081&m=dev)

**Summary**

- Add new Toast component using radix
- Needed for several client portal actions, like favoriting, to provide success or failure messages to the user

**Change List (describe the changes made to the files)**

This pull request introduces a new `Toast` component for displaying non-disruptive notifications, along with its supporting context, styles, and tests. The changes include the addition of the `Toast` component, a context provider for managing toast notifications, utility hooks, and comprehensive tests and stories for the component.

### New `Toast` Component Implementation:

* **Toast Component**: Added a `Toast` component that provides a notification system with a title, close button, and optional action button or link. It leverages `@radix-ui/react-toast` for functionality and includes custom styles. [[1]](diffhunk://#diff-a0f14f94bdb6e0ef299ff343c7f8283740406bc31c710fa9553aed23d068b506R1-R54) [[2]](diffhunk://#diff-ad2372394221b09e685e404a098c529281d16b34dad8960ab72b7964e4fb0487R1-R99)
* **Toast Context Provider**: Introduced `ToastProvider` to manage the state of toasts, allowing for adding and removing toast notifications programmatically.
* **Custom Hook**: Added `useToast` to simplify triggering toast notifications from any component.

### Supporting Changes:

* **Styles**: Added a SCSS file for the `Toast` component, including animations and responsive styling for the toast viewport.
* **Exports**: Updated `index.ts` to export the `Toast` component and its related utilities. [[1]](diffhunk://#diff-f2292efc82c7ad90272ff68ec13da542e3329805cbfcfd922e5734b498fa0f39R1-R3) [[2]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R101)

### Testing and Documentation:

* **Tests**: Comprehensive unit tests were added for the `Toast` component, `ToastProvider`, and `useToast` hook to ensure functionality and edge case handling. [[1]](diffhunk://#diff-7d8a553694af066f78c3a9d0b31cbe0591327dea0e443eb9c684825d6ba0323dR1-R86) [[2]](diffhunk://#diff-db6554c92ada48e20953f58b75dd1ddac74ab66add3adb5c0a7107304fb5f5aaR1-R139) [[3]](diffhunk://#diff-2adcb6c571495719d00737c3051fd0c63059e1636b2bb2715664ba7880e188f4R1-R74)
* **Stories**: Added Storybook stories for static and interactive demonstrations of the `Toast` component.

### Dependency and File Updates:

* **Dependencies**: Added `@radix-ui/react-toast` to `package.json` for toast functionality.
* **File Imports**: Updated `src/componentStyles.scss` to include the `Toast` component styles.

**Acceptance Test (how to verify the PR)**

- Add step by step instructions on how to verify the change

**Regression Test**

- (Optional) Add verification steps to make sure this PR doesn't break old functionality

**Evidence of testing**

- Post logs, screenshots, etc

**Things to look for during review**

- [ ] PR title should correctly describe the most significant type of commit. I.e. `feat(scope): ...` if a `minor` release should be triggered.
- [ ] All commit messages follow convention and are appropriate for the changes
- [ ] All references to `phillips` class prefix are using the prefix variable
- [ ] All major areas have a `data-testid` attribute.
- [ ] Document all props with jsdoc comments
- [ ] All strings should be translatable.
- [ ] Unit tests should be written and should have a coverage of 90% or higher in all areas.

New Components

- [ ] Are there any [accessibility considerations](https://www.w3.org/WAI/ARIA/apg/patterns/) that need to be taken into account and tested?
- [ ] Default story called "Playground" should be created for all new components
- [ ] Create a jsdoc comment that has an Overview section and a link to the Figma design for the component
- [ ] Export the component and its typescript type from the `index.ts` file
- [ ] Import the component scss file into the `componentStyles.scss` file.

[L3-6424]: https://phillipsauctions.atlassian.net/browse/L3-6424?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/Toast/ToastContextProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * Provider component that manages toast notifications state and operations
+ */
+export const ToastProvider = ({ children }: PropsWithChildren) => {
+  const [toasts, setToasts] = useState<Toast[]>([]);
+
+  /**
+   * Adds a new toast notification
+   */
+  const addToast = useCallback((toast: Omit<Toast, 'id'>) => {
+    const id = Math.random().toString();
```
[the remark this answers is not in the document]
Comment (DaveMcNamara) [reply src/components/Toast/ToastContextProvider.tsx:33]: `randomUUID` from the crypto library would be better, either way it's ridiculously unlikely to cause a collision.

File: src/components/Toast/ToastContextProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+interface Toast extends ToastPrimitive.ToastProps {
+  id: string;
+  title?: string;
+  description?: string;
+}
+
+export interface ToastContextType {
+  toasts: Toast[];
+  addToast: (toast: Omit<Toast, 'id'>) => void;
```
Comment (DaveMcNamara) [inline src/components/Toast/ToastContextProvider.tsx:14]: Omitting ID so that components using the hook don't have to worry about supplying one to this function?

File: src/components/Toast/Toast.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+const Toast = forwardRef<HTMLLIElement, PrimitiveToastProps>(({ className, title, action, ...props }, ref) => {
+  const { className: baseClassName, ...commonProps } = getCommonProps(props, 'Toast');
+
+  return (
+    <PrimitiveToast.Root {...commonProps} className={classnames(baseClassName, className)} ref={ref}>
+      <div className={`${baseClassName}__content`}>
+        <PrimitiveToast.Title>{title}</PrimitiveToast.Title>
+        {action && (
+          <PrimitiveToast.Action asChild altText="Toast action" className={`${baseClassName}__action`}>
```
Comment (DaveMcNamara) [inline src/components/Toast/Toast.tsx:38]: If this alt text is going to be visible/accessible we should probably either reference the title or let the caller supply a label.

File: src/components/Toast/ToastContextProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import * as ToastPrimitive from '@radix-ui/react-toast';
+import { type PropsWithChildren, createContext, useCallback, useState } from 'react';
+import Toast from './Toast';
+import { px } from '../../utils';
+
+interface Toast extends ToastPrimitive.ToastProps {
+  id: string;
+  title?: string;
```
Comment (DaveMcNamara) [inline src/components/Toast/ToastContextProvider.tsx:8]: Is there a case where we wouldn't need a title? Just want to make sure that optional properties are truly optional.

File: src/components/Toast/Toast.test.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import { render, screen } from '@testing-library/react';
+import { describe, it, expect, vi } from 'vitest';
+import { ToastProvider } from '@radix-ui/react-toast';
```
Comment (DaveMcNamara) [inline src/components/Toast/Toast.test.tsx:3]: Not the one from ToastContextProvider.tsx?

File: src/components/Toast/Toast.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import classnames from 'classnames';
+import * as PrimitiveToast from '@radix-ui/react-toast';
+import { Icon } from '../Icon';
+import IconButton from '../IconButton/IconButton';
+import { ButtonVariants } from '../Button/types';
+
+export interface PrimitiveToastProps extends PrimitiveToast.ToastProps {
+  title: string;
+  action?: ReactNode;
+}
```
Comment (DaveMcNamara) [inline src/components/Toast/Toast.tsx:12]: extends `ToastOptions` from the hook file instead of this object literal?

File: src/components/Toast/_toast.scss [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+.#{$px}-toast {
+  align-items: center;
+  background-color: $primary-black;
+  border-radius: 4px;
+  box-shadow: 0 6px 14px 0 $primary-black;
+  color: $white;
+  column-gap: $spacing-sm;
+  display: grid;
+  grid-template-areas: 'title action' 'description action';
```
Comment (DaveMcNamara) [inline src/components/Toast/_toast.scss:11]: I <3 grid template area names

File: src/components/Toast/_toast.scss [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  to {
+    transform: translateX(calc(100% + #{$spacing-lg}));
+  }
+}
+
+.#{$px}-toast-viewport {
+  --viewport-padding: #{$spacing-lg};
+
```
Comment (DaveMcNamara) [inline src/components/Toast/_toast.scss:86]: Can these two lines go or are they hanging out for a future reason?

File: src/components/Toast/useToast.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { ReactNode } from 'react';
+import { useToastContext } from './useToastContext';
+
+interface ToastOptions {
+  title: string;
+  action?: ReactNode;
```
Comment (DaveMcNamara) [inline src/components/Toast/useToast.tsx:6]: "action" sounds like it wants a function, not a component (which are functions but y'know). Maybe a better name would be "actionComponent"? Or separate props for `action` and `actionComponent` with the component defaulting to `button`? I don't know, I was just surprised that `action` wasn't roughly the same as `onClick`.

File: src/components/Toast/useToast.test.tsx [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+describe('useToast', () => {
+  it('shows toast with basic config', () => {
+    render(
+      <ToastProvider>
+        <TestComponent />
+      </ToastProvider>,
+    );
+
+    act(() => {
```
Comment (DaveMcNamara) [inline src/components/Toast/useToast.test.tsx:34]: My understanding is that `act` should rarely be necessary. When Vitest complains that "an action was not wrapped in act" what it means is "something changed that I didn't expect!"—which is typically solved by making the test async and using `findBy` and `waitFor` functions instead of `getBy`. Apologies if I'm off base with this diagnosis.

**Review by DaveMcNamara — CHANGES_REQUESTED** [review]
Had more to say than I expected! Overall this component looks excellent, none of my comments are blockers but I'm requesting changes anyway because certainly at least one of them warrants attention.

File: src/components/Toast/_toast.scss [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+  border-radius: 4px;
+  box-shadow: 0 6px 14px 0 $primary-black;
+  color: $white;
+  column-gap: $spacing-sm;
+  display: grid;
+  grid-template-areas: 'title action';
+  grid-template-columns: auto max-content;
+
+  // There are no spacing tokens for the padding used in figma
+  padding: 6px 16px;
```
Comment (DaveMcNamara) [inline src/components/Toast/_toast.scss:15]: `$spacing-xsm $spacing-small` would be pretty darn close (but reponsive, so slightly shifting). Not sure that this is better for this use case, but wanted to mention!

**Review by DaveMcNamara — APPROVED** [review]

File: src/components/Toast/Toast.tsx [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+          <PrimitiveToast.Title>{title}</PrimitiveToast.Title>
+          {actionElement && actionAltText && (
+            <PrimitiveToast.Action asChild altText={actionAltText} className={`${baseClassName}__action`}>
+              {actionElement}
+            </PrimitiveToast.Action>
+          )}
+        </div>
+        <PrimitiveToast.Close className={`${baseClassName}__close`} aria-label="Close" asChild>
+          <IconButton variant={ButtonVariants.link}>
+            <Icon icon="CloseX" title="Close" color="$white" />
```
Comment (scottdickerson) [inline src/components/Toast/Toast.tsx:47]: "Close" needs to be translated

File: src/components/Toast/Toast.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+import Toast from './Toast';
+import { ToastProvider } from './ToastContextProvider';
+import Button from '../Button/Button';
+import { useToast } from './useToast';
+import { ButtonVariants } from '../Button/types';
+
+const meta = {
+  title: 'Components/Toast',
+  component: Toast,
+  decorators: [(Story) => <ToastProvider>{Story()}</ToastProvider>],
```
Comment (scottdickerson) [inline src/components/Toast/Toast.stories.tsx:11]: should this be wrapped globally in case our other components need it in their stories?

File: src/components/Toast/_toast.scss [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    transform: translateX(calc(100% + #{$spacing-lg}));
+  }
+}
+
+.#{$px}-toast-viewport {
+  bottom: 0;
+  display: flex;
+  flex-direction: column;
+  gap: $spacing-sm;
+  left: 0;
```
[the remark this answers is not in the document]
Comment (scottdickerson) [reply src/components/Toast/_toast.scss:89]: I'm fine with opening this as a follow-on issue if the error is major, I just wanna keep track of these ideas somewhere

**Review by scottdickerson — CHANGES_REQUESTED** [review]
Some additional discussions to resolve here

File: src/components/Toast/Toast.tsx [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import classnames from 'classnames';
+import * as PrimitiveToast from '@radix-ui/react-toast';
+import { Icon } from '../Icon';
+import IconButton from '../IconButton/IconButton';
+import { ButtonVariants } from '../Button/types';
+
+export interface PrimitiveToastProps extends PrimitiveToast.ToastProps {
+  title: string;
+  actionElement?: ReactNode;
+  actionAltText?: string;
```
Comment (davidicus) [inline src/components/Toast/Toast.tsx:12]: For consistency can we add comments here for the storybook documentation?

**Review by davidicus — COMMENTED** [review]
While the toast auto dismissing may be the current behavior, do we want to add a prop to control this? For a11y reasons, I do not think we should be auto dismissing our toast.

**Review by scottdickerson — APPROVED** [review]
